File: docs/algebra/fast-fourier-transform/src/LG-P3803-Iteration.kt
import kotlin.math.*

fun readInts(): List<Int> {
    return readLine()!!.trim().split(" ").map(String::toInt)
}

fun nearest(n: Int): Int {
    return 1.shl(ceil(log2(n.toDouble())).toInt() + 1)
}

data class Complex(val real: Double, val imag: Double) {
    operator fun plus(other: Complex): Complex {
        return Complex(real + other.real, imag + other.imag)
    }

    operator fun minus(other: Complex): Complex {
        return Complex(real - other.real, imag - other.imag)
    }

    operator fun times(other: Double): Complex {
        return Complex(real * other, imag * other)
    }

    operator fun times(other: Complex): Complex {
        return Complex(real * other.real - imag * other.imag, real * other.imag + imag * other.real)
    }
}

val rev = IntArray(nearest(1000002))

fun change(f: Array<Complex>, n: Int) {
    if (rev[1] == 0)
        for (i in 0 until n) {
            rev[i] = rev[i shr 1] shr 1
            if (i % 2 == 1)
                rev[i] = rev[i] or (n shr 1)
        }
    for (i in 0 until n)
        if (i < rev[i]) {
            val tmp = f[i]
            f[i] = f[rev[i]]
            f[rev[i]] = tmp
        }
}

fun fft(f: Array<Complex>, n: Int, rev: Int) {
    change(f, n)
    var len = 2
    while (len <= n) {
        val omega = Complex(cos(2.0 * PI / len * rev), sin(2.0 * PI / len * rev))
        var l = 0
        while (l < n) {
            var now = Complex(1.0, 0.0)
            for (i in l until l + len / 2) {
                val g = f[i]
                val h = now * f[i + len / 2]
                f[i] = g + h
                f[i + len / 2] = g - h
                now *= omega
            }
            l += len
        }
        len *= 2
    }
    if (rev == -1)
        for (i in 0 until n)
            f[i] = f[i] * (1.0 / n)
}

fun main() {
    val (n, m) = readInts()
    val k = nearest(n + m + 1)
    val a = Array(k) { Complex(0.0, 0.0) }
    for ((i, v) in readInts().withIndex())
        a[i] = Complex(v.toDouble(), 0.0)
    val b = Array(k) { Complex(0.0, 0.0) }
    for ((i, v) in readInts().withIndex())
        b[i] = Complex(v.toDouble(), 0.0)
    fft(a, k, 1)
    fft(b, k, 1)
    for (i in 0 until k)
        a[i] = a[i] * b[i]
    fft(a, k, -1)
    val ans = a.slice(0 until n + m + 1).map { round(it.real).toInt() }
    println(ans.joinToString(" "))
}

File: docs/algebra/fast-fourier-transform/src/LG-P3803-Iteration.cpp
#include <cmath>
#include <complex>
#include <iostream>
#define MAXN (1 << 22)

using namespace std;
typedef complex<double> cd;
const cd I{0, 1};
cd a[MAXN], b[MAXN];
void change(cd *f, int n) {
  int i, j, k;
  for (int i = 1, j = n / 2; i < n - 1; i++) {
    if (i < j)
      swap(f[i], f[j]);
    k = n / 2;
    while (j >= k) {
      j = j - k;
      k = k / 2;
    }
    if (j < k)
      j += k;
  }
}
void fft(cd *f, int n, int rev) {
  change(f, n);
  for (int len = 2; len <= n; len <<= 1) {
    cd omega = exp(I * (2 * M_PI / len * rev));
    for (int j = 0; j < n; j += len) {
      cd now = 1;
      for (int k = j; k < j + len / 2; ++k) {
        cd g = f[k], h = now * f[k + len / 2];
        f[k] = g + h, f[k + len / 2] = g - h;
        now *= omega;
      }
    }
  }
  if (rev == -1)
    for (int i = 0; i < n; ++i)
      f[i] /= n;
}
int main() {
  int n, m;
  cin >> n >> m;
  int k = 1 << (32 - __builtin_clz(n + m + 1));
  for (int i = 0; i <= n; ++i)
    cin >> a[i];
  for (int j = 0; j <= m; ++j)
    cin >> b[j];
  fft(a, k, 1);
  fft(b, k, 1);
  for (int i = 0; i < k; ++i)
    a[i] *= b[i];
  fft(a, k, -1);
  for (int i = 0; i < n + m + 1; ++i)
    cout << (int)round(a[i].real()) << " ";
}
